Use SingleFileHost

swaroop-sridhar opened the pull request:
This commit implements the following changes for single-file apps:
* When publishing self-contained single-file apps:
  * Use `SingleFileHost` instead of `apphost`
  * Trim the native components of the runtime published for the app.
   * Fixes #11567.
* Implements the [optional additional settings](https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#optional-settings) for .net 5 `PublishSingleFile`.

swaroop-sridhar committed:
Use SingleFileHost

This commit implements the following changes for single-file apps:
* When publishing self-contained single-file apps:
  * Use `SingleFileHost` instead of `apphost`
  * Trim the native components of the runtime published for the app.
   * Fixes #11567.

* Implements the [optional additional settings](https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#optional-settings) for .net 5 `PublishSingleFile`.
  * https://github.com/dotnet/runtime/issues/36590

swaroop-sridhar commented on src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishASingleFileApp.cs:
```diff
@@ -23,13 +23,17 @@ public class GivenThatWeWantToPublishASingleFileApp : SdkTest
+        private const string IncludePdb = "/p:IncludeSymbolsInSingleFile=true";
+        private const string IncludeNative = "/p:IncludeNativeLibrariesInSingleFile=true";
+        private const string IncludeAllContent = "/p:IncludeAllContentInSingleFile=true";
+        private const string ExcludeContent = "/p:IncludeAllContentInSingleFile=false";
```
The `ExcludeContent` flag is only needed until Runtime changes to process assemblies from bundle is available in the installer repo, so that `global.json` can be updated to use the new runtime. Once this is done, the default will be set to `IncludeAllContentInSingleFile=false` in the publish-targets.

swaroop-sridhar commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -1002,7 +1014,9 @@ Copyright (c) .NET Foundation. All rights reserved.
       <IntermediateDepsFilePath Condition=" '$(PublishDepsFilePath)' == ''">$(IntermediateOutputPath)$(ProjectDepsFileName)</IntermediateDepsFilePath >
-      <PublishDepsFilePath Condition=" '$(PublishDepsFilePath)' == '' And '$(PublishSingleFile)' != 'true'">$(PublishDir)$(ProjectDepsFileName)</PublishDepsFilePath>      
+      <PublishDepsFilePath Condition=" '$(PublishDepsFilePath)' == '' And '$(PublishSingleFile)' != 'true'">$(PublishDir)$(ProjectDepsFileName)</PublishDepsFilePath>
+      <_IsSingleFilePublish Condition="'$(PublishSingleFile)' == ''">false</_IsSingleFilePublish>
+      <_IsSingleFilePublish Condition="'$(PublishSingleFile)' != ''">$(PublishSingleFile)</_IsSingleFilePublish>
```
The `PublishSingleFile` property is captured by telemetry, which reports the three states: `true`, `false`, and `unset`. So, I wanted to leave the user input in its pure state and use a different variable.

swaroop-sridhar committed:
Address feedback from @dplaisted.

swaroop-sridhar commented:
Thanks for the review, @dsplaisted. I have committed the changes you've suggested.

swaroop-sridhar commented:
The Linux/Osx failures are because `singelfilehost` was not packaged into the host packages for these platforms. These will get fixed once https://github.com/dotnet/runtime/pull/37147 is available in the installer repo.

dsplaisted commented:
> The Linux/Osx failures are because `singelfilehost` was not packaged into the host packages for these platforms. These will get fixed once [dotnet/runtime#37147](https://github.com/dotnet/runtime/pull/37147) is available in the installer repo.

So once that is fixed then you should be able to update stage 0 for this PR and merge it?

swaroop-sridhar commented:
> So once that is fixed then you should be able to update stage 0 for this PR and merge it?

Yes that is my plan -- would like to have it in Preview 6.
* The fix for `singlefilehost` is already checked in the runtime repo. I'm waiting for the changes to propagate to SDK->Installer repos, so that I can update stage0.
* With a more recent SDK (5.0.100-preview.6.20278.11) there's a new unrelated failure on Windows. Looks like there's a bug in the crossgen2 package (missing clrcompression.dll). I'm working on getting that fixed.

swaroop-sridhar committed:
Update global.json

(To include update from runtime)

swaroop-sridhar committed:
Try disabling a crossgen2 test

Updating stage0 caused crossgen2 tests to fail because of missing dependencies in the corssgen2 package.
https://github.com/dotnet/runtime/issues/37196

Therefore, try disabling the tests temporarily.

swaroop-sridhar commented:
@dsplaisted  Updating stage0 triggered crossgen2  failure: dotnet/runtime#37196.
Therefore, I disabled two crossgen2 tests in this commit: d48f3e0

I think it is a good idea to disable these tests temporarily, rather than holding the global.json update (and this PR) until dotnet/runtime#37196 is fixed. 
I wanted to call this out before checking in. Thanks.

CC: @trylek.

dsplaisted commented:
> @dsplaisted Updating stage0 triggered crossgen2 failure: [dotnet/runtime#37196](https://github.com/dotnet/runtime/issues/37196).
> Therefore, I disabled two crossgen2 tests in this commit: [d48f3e0](https://github.com/dotnet/sdk/commit/d48f3e00b3daf3c9105195a394c17c4a7584c1b2)
> 
> I think it is a good idea to disable these tests temporarily, rather than holding the global.json update (and this PR) until [dotnet/runtime#37196](https://github.com/dotnet/runtime/issues/37196) is fixed.
> I wanted to call this out before checking in. Thanks.
> 
> CC: @trylek.

Sounds good.  We don't have a great process to track re-enabling tests like this.  Will you remember to re-enable these?  We could also file an issue in the sdk repo to track re-enabling the tests when the bug is fixed.

swaroop-sridhar commented:
OK @dsplaisted, I filed https://github.com/dotnet/sdk/issues/11845 to re-enable the tests once the underlying issue is fixed.
